docs/madr/decisions/033-meshhttproute-meshgateway.md

```diff
@@ -0,0 +1,133 @@
+# `MeshHTTPRoute`/`MeshTCPRoute` for `MeshGateway`
+
+- Status: accepted
+
+Technical Story: #8142
+
+## Context and Problem Statement
+
+We have some very similar resources, `MeshHTTPRoute`/`MeshTCPRoute` and `MeshGatewayRoute` that
+both configure route matching, filtering and redirecting semantics.
+This MADR proposes adapting `MeshHTTPRoute` and `MeshTCPRoute` for use with
+`MeshGateway.`
+
+## Decision Drivers <!-- optional -->
+
+* Fewer CRDs
+* Parity with `MeshGatewayRoute`
+
+## Considered Options
+
+- Keep two resources
+- Use `MeshHTTPRoute`/`MeshTCPRoute`
+
+## Decision Outcome
+
+Chosen option:
+
+- Adapt `MeshHTTPRoute`/`MeshTCPRoute`
+- Add `spec.to[].hostnames` to `MeshHTTPRoute`
+- Allow `spec.to[].targetRef` to be empty
+- Require `backendRefs` to be set
+- Don't support policy matching on `MeshHTTPRoute`/`MeshTCPRoute` if it attaches to a gateway
+
+### Hostname
+
+With `MeshGatewayRoute` and `HTTPRoute`, users can set hostnames to only apply
+the route rules to requests going to the given hosts.
+We don't have this possibility with `MeshHTTPRoute`, so we can either:
+
+- add `hostnames` to the `MeshHTTPRoute`, likely directly under `spec.to`
+  entries
+- have users on each rule do hostname matching using the header match on `host`
+
+This MADR proposes adding `spec.to[].hostnames`. Hostname matching is in some sense the
+"highest level" matching possible with HTTP and it's likely that all rules of a
+given resource are intended to apply to a given hostname set, whether that's a
+specific hostname, a wildcard pattern or to all hostnames.
+
+The second option might be feasible but it only makes sense if the cost of having
+host matching separate is very high. It also differs from Gateway API, which is
+a disadvantage.
+
+### `to.targetRef`
+
+When using `MeshHTTPRoute`/`MeshTCPRoute` for service to service communication, it's very
+useful to limit the effects based on request source and destination. We do
+this by setting `spec.targetRef`, the source, and `spec.to[].targetRef`, the destination.
+
+With `MeshGateway`, once a request hits the gateway, the routing resources
+_define_ the destination. We have to put the rules somewhere, so this MADR
+proposes `spec.to` but there's no semantic value to a `spec.to[].targetRef`
+or for that matter, a `spec.from[].targetRef`.
+
+This MADR proposes allowing `spec.to[].targetRef` to be omitted. Another
+possibility would be requiring `spec.to[].targetRef.kind: Mesh` but it's
+not clear that this has any advantages.
+
+### Route matching
+
+This MADR doesn't address supporting policy matching of routes
+that attach to `MeshGateways` because we can't policy match on `MeshGatewayRoute` at
+the moment anyway. The feasability/details should be addressed in a separate MADR.
+
+### `MeshHTTPRoute` example
+
+```
+apiVersion: kuma.io/v1alpha1
```

Comment:
What backend will be picked if I send a request to `test.example.com`?
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: edge-routes-a
spec:
  targetRef:
    kind: MeshGateway
    name: edge-gateway
  to:
  - hostnames:
    - test.example.com
    rules:
    - matches:
      - path:
          type: PathPrefix
          value: /
      default:
        backendRefs:
        - kind: MeshService
          name: backend_a
---
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: edge-routes-b
spec:
  targetRef:
    kind: MeshGateway
    name: edge-gateway
  to:
  - hostnames:
    - *.example.com
    rules:
    - matches:
      - path:
          type: PathPrefix
          value: /
      default:
        backendRefs:
        - kind: MeshService
          name: backend_b
```

Reply:
The first.

Reply:
Semantically, they'd create a  set of rules that can only be merged with other rule sets that have the exact same `hostname`. I tend to think it's not feasible to do this across wildcards, only between exact matches.

My assumption here was "nothing new, it works just like MeshGatewayRoute." Which I think is still basically true but I'll make it explicit and write up some examples as well.

As far as the implementation, we'll see

Reply:
Yeah ideally I feel like this should just be similar to being part of `matches` section.

Reply:
Are `hostnames` basically a syntactic sugar for `matches`? For example
```yaml
to:
- hostnames:
  - test1.example.com
  - *.domain.com
  rules:
    - matches:
      - path:
          type: PathPrefix
          value: /
```
results in 
```yaml
to:
- rules:
    - matches:
      - path:
          type: PathPrefix
          value: /
      - headers:
        - type: Exact
          name: Host
          value: test1.example.com
        - type: RegularExpression
          name: Host
          value: *.domain.com
```

If we treat `hostnames` as syntactic sugar, we don't have to introduce additional complexity into the matching algorithm. Also, `hostnames` naturally becomes part of the `mergeKey`

Reply:
I can't come up with an example where it has different semantics.